### PR TITLE
Cleanup 0.15

### PR DESCRIPTION
**Pull request by vchuravy**
# TODO:
- [x] add tests for code_typed and code_warntype (including interactive=true)
- [x] Add code_warntype

### CODE REVIEW

File: src/reflection.jl [hunk 1]
```diff
@@ -39,17 +39,43 @@ function highlight(io::IO, code, lexer)
...
+    return map(Core.Compiler.specialize_method, method_matches(tt; world))
+end
 
 #
 # code_* replacements
 #
 
-code_lowered(@nospecialize(job::CompilerJob); kwargs...) =
-    InteractiveUtils.code_lowered(job.source.f, job.source.tt; kwargs...)
+function code_lowered(@nospecialize(job::CompilerJob); generated::Bool=true, debuginfo::Symbol=:default)
```
Comment (maleadt) [inline src/reflection.jl:59]: Ah, I had hoped we didn't have to replicate the entire definition...

Comment (vchuravy) [reply src/reflection.jl:59]: Yeah didn't see a way around it

**Comment by vchuravy** [comment]
cc: @wsmoses

File: src/reflection.jl (hunk 1 above)
Comment (maleadt) [reply src/reflection.jl:59]: Add a note then, we preferably want to extend InteractiveUtils not to need this def anymore.

**Comment by vchuravy** [comment]
@maleadt onw way of avoiding copying Base implementations would be to add a real dependency on Cthulhu.

File: src/reflection.jl [hunk 2]
```diff
@@ -39,51 +39,90 @@ function highlight(io::IO, code, lexer)
...
+    u = Base.unwrap_unionall(job.source.tt)
+    return Base.rewrap_unionall(Tuple{job.source.f, u.parameters...}, job.source.tt)
+end
+
+function method_instances(@nospecialize(tt::Type), world::UInt=Base.get_world_counter())
+    return map(Core.Compiler.specialize_method, method_matches(tt; world))
+end
+
+# FIXME: This is a horrible hack, is there a better way?
+function mi_to_func(mi::Core.MethodInstance)
```
Comment (vchuravy) [inline src/reflection.jl:56]: @vtjnash there is no better way for this right? We want to go from `Tuple{typeof(f), Int}` to `f, (Int,)`

Comment (vtjnash) [reply src/reflection.jl:56]: In general, the transform is unsound, so it doesn't exist. But there is tuple-type-tail perhaps

Comment (vchuravy) [reply src/reflection.jl:56]: Right but `Base.tuple_type_tail` and `Base.tuple_type_head` will return `typeof(+)` and not the instance `+` which reflection requires.

**Comment by vchuravy** [comment]
Sigh of course:

```
julia> GPUCompiler.code_typed(job; interactive=true)
ERROR: MethodError: no method matching descend_code_typed(::Type{Tuple{typeof(identity), Int64}}; interp=GPUCompiler.GPUInterpreter(GPUCompiler.CodeCache(Dict{Core.MethodInstance, Vector{Core.CodeInstance}}()), nothing, Core.Compiler.InferenceResult[], 0x0000000000007403, Core.Compiler.InferenceParams(true, false, false, 3, 4, 8, 3, 32), Core.Compiler.OptimizationParams(true, 100, 1000, 250, 20, 3, 32, 4, false)))
Closest candidates are:
  descend_code_typed(::Any, ::Any; kwargs...) at /home/vchuravy/.julia/packages/Cthulhu/wsvdB/src/Cthulhu.jl:99
  descend_code_typed(::Core.MethodInstance; kwargs...) at /home/vchuravy/.julia/packages/Cthulhu/wsvdB/src/Cthulhu.jl:145
  descend_code_typed(::Cthulhu.Bookmark; kw...) at /home/vchuravy/.julia/packages/Cthulhu/wsvdB/src/Cthulhu.jl:275
Stacktrace:
 [1] code_typed(job::CompilerJob; interactive::Bool, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ GPUCompiler ~/src/GPUCompiler/src/reflection.jl:69
 [2] top-level scope
   @ REPL[9]:1
```

**Pull request merged**